### PR TITLE
MF-183 bulk reservation creation

### PR DESCRIPTION
**Pull request by dashton82**

### CODE REVIEW

File: src/SFA.DAS.Reservations.Api/Controllers/ReservationsController.cs [hunk 1]
```diff
@@ -182,5 +185,31 @@ public async Task<IActionResult> Delete(Guid id)
...
+        public async Task<IActionResult> BulkCreate(long accountLegalEntityId, uint count)
+        {
+            try
+            {
+                var result = await _mediator.Send(new BulkCreateAccountReservationsCommand { AccountLegalEntityId = accountLegalEntityId, ReservationCount = count});
+                return Created("", result);
+            }
+            catch (ArgumentException argumentException)
+            {
+                Console.WriteLine(argumentException);
```
Comment (scottywakefield) [inline src/SFA.DAS.Reservations.Api/Controllers/ReservationsController.cs]: worth logging?

File: src/SFA.DAS.Reservations.Application.UnitTests/AccountReservation/Commands/BulkCreateReservations/WhenBulkCreatingAccountReservations.cs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+            //Act
+            await handler.Handle(command, CancellationToken.None);
+
+            //Assert
+            mockAccountReservationService.Verify(x=>x.BulkCreateAccountReservation(command.ReservationCount,command.AccountLegalEntityId, accountLegalEntity.AccountId, accountLegalEntity.AccountLegalEntityName));
+        }
+
+        [Test, MoqAutoData]
+        public async Task ThenWillReturnCreatedReservationIds(
```
Comment (scottywakefield) [inline src/SFA.DAS.Reservations.Application.UnitTests/AccountReservation/Commands/BulkCreateReservations/WhenBulkCreatingAccountReservations.cs]: inconsistent naming style

File: src/SFA.DAS.Reservations.Application.UnitTests/AccountReservation/Commands/BulkCreateReservations/WhenBulkCreatingAccountReservations.cs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+            //Act
+            var result = await handler.Handle(command, CancellationToken.None);
+
+            //Assert
+            Assert.AreEqual(createdReservationIds, result.ReservationIds);
+        }
+
+        [Test, MoqAutoData]
+        public void ThenWillThrowServiceExceptions(
```
Comment (scottywakefield) [inline src/SFA.DAS.Reservations.Application.UnitTests/AccountReservation/Commands/BulkCreateReservations/WhenBulkCreatingAccountReservations.cs]: inconsistent naming style

File: src/SFA.DAS.Reservations.Application.UnitTests/AccountReservation/Commands/BulkCreateReservations/WhenValidatingBulkCreateAccountReservations.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
+﻿using System.Threading.Tasks;
+using NUnit.Framework;
+using SFA.DAS.Reservations.Application.AccountReservations.Commands.BulkCreateAccountReservations;
+
+namespace SFA.DAS.Reservations.Application.UnitTests.AccountReservation.Commands.BulkCreateReservations
+{
+    public class WhenValidatingBulkCreateAccountReservations
+    {
+        [Test]
+        public async Task ThenAValidCommandShouldPass()
```
Comment (scottywakefield) [inline src/SFA.DAS.Reservations.Application.UnitTests/AccountReservation/Commands/BulkCreateReservations/WhenValidatingBulkCreateAccountReservations.cs]: more inconsistent names in this file

File: src/SFA.DAS.Reservations.Application.UnitTests/AccountReservation/Commands/BulkCreateReservations/WhenValidatingBulkCreateAccountReservations.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            var result = await validator.ValidateAsync(command);
+
+            //Assert
+            Assert.IsFalse(result.IsValid());
+            Assert.AreEqual(1, result.ValidationDictionary.Count);
+            Assert.IsTrue(result.ValidationDictionary.ContainsKey(nameof(command.ReservationCount)));
+        }
+
+        [Test]
+        public async Task ThenIfReservationCountIsSetToInvalidValuesShouldFail()
```
Comment (scottywakefield) [inline src/SFA.DAS.Reservations.Application.UnitTests/AccountReservation/Commands/BulkCreateReservations/WhenValidatingBulkCreateAccountReservations.cs]: this test is the same as the test above.

File: src/SFA.DAS.Reservations.Domain/Reservations/Reservation.cs [hunk 6]
```diff
@@ -79,8 +79,13 @@ public Reservation(Guid id, long accountId, DateTime startDate, int expiryPeriod
 
         private IList<Rule> GetRules(Func<DateTime, Task<IList<Rule>>> getRules)
         {
-            var task = getRules(StartDate);
-            return task.Result;
+            if (StartDate.HasValue)
```
Comment (scottywakefield) [inline src/SFA.DAS.Reservations.Domain/Reservations/Reservation.cs]: I can't find any tests for this if scenario, e.g. if no value, if has value

File: src/SFA.DAS.Reservations.Domain/Reservations/Reservation.cs [hunk 7]
```diff
@@ -104,11 +109,16 @@ private static ApprenticeshipCourse.Course MapCourse(Course reservationCourse)
             return reservationCourse == null ? null : new ApprenticeshipCourse.Course(reservationCourse);
         }
 		
-		private DateTime GetExpiryDateFromStartDate(int expiryPeriodInMonths)
+		private DateTime? GetExpiryDateFromStartDate(int expiryPeriodInMonths)
         {
-            var expiryDate = StartDate.AddMonths(expiryPeriodInMonths);
-            var lastDayInMonth = DateTime.DaysInMonth(expiryDate.Year, expiryDate.Month);
-            return new DateTime(expiryDate.Year, expiryDate.Month, lastDayInMonth);
+            if (StartDate.HasValue)
```
Comment (scottywakefield) [inline src/SFA.DAS.Reservations.Domain/Reservations/Reservation.cs]: I can't find any tests for this if scenario, e.g. if no value, if has value

File: src/SFA.DAS.Reservations.Api/Controllers/ReservationsController.cs (hunk 1 above)
Comment (dashton82) [reply src/SFA.DAS.Reservations.Api/Controllers/ReservationsController.cs]: Updated all console logs

File: src/SFA.DAS.Reservations.Application.UnitTests/AccountReservation/Commands/BulkCreateReservations/WhenBulkCreatingAccountReservations.cs (hunk 2 above)
Comment (dashton82) [reply src/SFA.DAS.Reservations.Application.UnitTests/AccountReservation/Commands/BulkCreateReservations/WhenBulkCreatingAccountReservations.cs]: Changed

File: src/SFA.DAS.Reservations.Application.UnitTests/AccountReservation/Commands/BulkCreateReservations/WhenBulkCreatingAccountReservations.cs (hunk 3 above)
Comment (dashton82) [reply src/SFA.DAS.Reservations.Application.UnitTests/AccountReservation/Commands/BulkCreateReservations/WhenBulkCreatingAccountReservations.cs]: Changed

File: src/SFA.DAS.Reservations.Application.UnitTests/AccountReservation/Commands/BulkCreateReservations/WhenValidatingBulkCreateAccountReservations.cs (hunk 4 above)
Comment (dashton82) [reply src/SFA.DAS.Reservations.Application.UnitTests/AccountReservation/Commands/BulkCreateReservations/WhenValidatingBulkCreateAccountReservations.cs]: Changed

File: src/SFA.DAS.Reservations.Application.UnitTests/AccountReservation/Commands/BulkCreateReservations/WhenValidatingBulkCreateAccountReservations.cs (hunk 5 above)
Comment (dashton82) [reply src/SFA.DAS.Reservations.Application.UnitTests/AccountReservation/Commands/BulkCreateReservations/WhenValidatingBulkCreateAccountReservations.cs]: Removed test

File: src/SFA.DAS.Reservations.Domain/Reservations/Reservation.cs (hunk 6 above)
Comment (dashton82) [reply src/SFA.DAS.Reservations.Domain/Reservations/Reservation.cs]: Added tests to cover this scenario

**Pull request merged**